sensing/position_estimation.c

```diff
@@ -212,49 +209,60 @@ static void position_estimation_position_correction(position_estimation_t *pos_e
 			local_coordinates = coord_conventions_global_to_local_position(global_gps_position,pos_est->local_position.origin);
 			local_coordinates.timestamp_ms = pos_est->gps->time_last_msg;
 			
-			// compute GPS velocity estimate
-			gps_dt = (local_coordinates.timestamp_ms - pos_est->last_gps_pos.timestamp_ms) / 1000.0f;
-			if (gps_dt > 0.001f)
-			{
-				for (i = 0; i < 3; i++)
-				{
-					pos_est->last_vel[i] = (local_coordinates.pos[i] - pos_est->last_gps_pos.pos[i]) / gps_dt;
-				}
-				pos_est->last_gps_pos = local_coordinates;
-			}
-			else
-			{
-				print_util_dbg_print("GPS dt is too small!\r\n");
-			}
+			pos_est->last_gps_pos = local_coordinates;
 		}
-		t_inter_gps = time_keeper_get_millis() - pos_est->gps->time_last_msg;
-			
-		//gps_gain = maths_f_max(1.0f - t_inter_gps / 1000.0f, 0.0f);
+		
+		vel_error[X] = pos_est->gps->north_speed    - pos_est->vel[X]; 
+		vel_error[Y] = pos_est->gps->east_speed     - pos_est->vel[Y]; 
+		vel_error[Z] = pos_est->gps->vertical_speed - pos_est->vel[Z]; 
+
 		gps_gain = 1.0f;
-			
+		
 		for (i = 0;i < 3;i++)
 		{
 			pos_error[i] = pos_est->last_gps_pos.pos[i] - pos_est->local_position.pos[i];
-			vel_error[i] = pos_est->last_vel[i]       - pos_est->vel[i]; 
 		}
 	}
 	else
 	{
 		gps_position_init(pos_est);
-		for (i = 0;i < 2;i++)
+		for (i = 0;i < 3;i++)
 		{
 			pos_error[i] = 0.0f;
 			vel_error[i] = 0.0f;
 		}
-		gps_gain = 0.1f;
+		gps_gain = 0.0f;
 	}
-		
+	
+	if (pos_est->sonar->healthy)
+	{
+		sonar_alt_error = -pos_est->sonar->current_distance - pos_est->local_position.pos[Z];
+
+		sonar_gain = 1.0f;
+	}
+	else
+	{
+		sonar_alt_error = 0.0f;
+
+		sonar_gain = 0.0f;
+	}
+
+	if (pos_est->sonar->healthy_vel)
+	{
+		sonar_vel_error = -pos_est->sonar->current_velocity - pos_est->vel[Z];
+	}
+	else
+	{
+		sonar_vel_error = 0.0f;
+	}
+
 	// Apply error correction to velocity and position estimates
 	for (i = 0;i < 3;i++)
 	{
 		pos_est->local_position.pos[i] += pos_est->kp_pos_gps[i] * gps_gain * pos_error[i]* dt;
 	}
-	pos_est->local_position.pos[2] += pos_est->kp_alt_baro * baro_gain * baro_alt_error* dt;
+	pos_est->local_position.pos[Z] += pos_est->kp_alt_baro * baro_gain * baro_alt_error* dt;
+	pos_est->local_position.pos[Z] += pos_est->kp_alt_sonar * sonar_gain * sonar_alt_error* dt;
```

Comment:
@ndousse 
why in the following for loop we do that copy of vel_error to vel_correction.v ?
Shouldn't we clean it, removing that useless vel_correction object ?